sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MobileDatabaseEncryptionKeysCheck.kt

```diff
@@ -0,0 +1,94 @@
+        FunMatcher(qualifier = SQLITE) {
+            withNames(CHANGE_PASSWORD, "openDatabase", "openOrCreateDatabase", "create")
+        },
+        FunMatcher(qualifier = "io.realm.RealmConfiguration.Builder", name = ENCRYPTION_KEY)
+    )
+
+    override fun visitFunctionCall(
+        callExpression: KtCallExpression,
+        resolvedCall: ResolvedCall<*>,
+        kotlinFileContext: KotlinFileContext
+    ) {
+        val bindingContext = kotlinFileContext.bindingContext
+        val functionName = resolvedCall.resultingDescriptor.name.asString()
+
+        val valueArgumentsList = resolvedCall.valueArgumentsByIndex
+        if (valueArgumentsList == null || (valueArgumentsList.size < 2 && functionName != ENCRYPTION_KEY)) return
+
+        val arg = if (functionName in setOf(ENCRYPTION_KEY, CHANGE_PASSWORD)) {
+            valueArgumentsList[0]
+        } else valueArgumentsList[1]
+        val argExpr = (arg as? ExpressionValueArgument)?.valueArgument?.getArgumentExpression() ?: return
+
+        val secondaries = mutableListOf<PsiElement>()
+        val argValueExpr = argExpr.predictRuntimeValueExpression(bindingContext, secondaries)
+        if (argValueExpr.isHardCoded(bindingContext, secondaries)) {
+            val parameter = if (functionName == ENCRYPTION_KEY) ENCRYPTION_KEY else "password"
+
+            kotlinFileContext.reportIssue(
+                callExpression.calleeExpression!!,
+                """The "$parameter" parameter should not be hardcoded.""",
+                secondaries.map { SecondaryLocation(kotlinFileContext.textRange(it)) },
+            )
+        }
+    }
+}
+
+private fun KtExpression.isHardCoded(bindingContext: BindingContext, secondaries: MutableList<PsiElement>) : Boolean =
+    when(this) {
+        is KtStringTemplateExpression -> true
+        is KtDotQualifiedExpression -> (selectorExpression?.isHardCoded(bindingContext, secondaries) ?: false)
+            || receiverExpression
+            .predictRuntimeValueExpression(bindingContext, secondaries)
+            .isHardCoded(bindingContext, secondaries)
+        is KtCallExpression -> CREATE_CHAR_BYTE_ARRAY.matches(this, bindingContext)
+            || returnsHardcoded(bindingContext, secondaries)
+        else -> false
+    }
+
+fun KtCallExpression.returnsHardcoded(bindingContext: BindingContext, secondaries: MutableList<PsiElement>) : Boolean {
+    val resultingDescriptor = this.getResolvedCall(bindingContext)?.resultingDescriptor ?: return false
+    val declaration = DescriptorToSourceUtils.descriptorToDeclaration(resultingDescriptor) as? KtNamedFunction ?: return false
+
+    if (!declaration.hasBody()) return false
+    if (declaration.hasBlockBody()) {
+        return declaration.anyDescendantOfType<KtReturnExpression> {
+            it.returnedExpression?.isHardCoded(bindingContext, secondaries) ?: false
+        }
+    }
+
+    return declaration.bodyExpression?.isHardCoded(bindingContext, secondaries) ?: false
```

Comment:
To me this seems like a logical `else` case for the `if(declaration.hasBlockBody())` statement